cmd/zoekt-sourcegraph-indexserver/sg.go

```diff
@@ -244,58 +327,112 @@ func (s *sourcegraphClient) List(ctx context.Context, indexed []uint32) (*Source
 		}
 		next += time.Duration(rand.Int63n(int64(next) / 4)) // jitter
 		s.configFingerprintReset = time.Now().Add(next)
+
+		s.configFingerprintProto = nil
 		s.configFingerprint = ""
 	}
 
-	// We want to use a consistent fingerprint for each call. Next time list is
-	// called we want to use the first fingerprint returned from the
-	// configuration endpoint. However, if any of our configuration calls fail,
-	// we need to fallback to our last value.
-	lastFingerprint := s.configFingerprint
-	first := true
+	// getIndexOptionsFunc is a function that can be used to get the index
+	// options for a set of repos (while properly handling any configuration fingerprint
+	// changes).
+	//
+	// In general, this function provides a consistent fingerprint for each batch call,
+	// and updates the server state with the new fingerprint. If any of the batch calls
+	// fail, the old fingerprint is restored.
+	type getIndexOptionsFunc func(repos ...uint32) ([]indexOptionsItem, error)
+
+	// default to REST
+	mkGetIndexOptionsFunc := func(tr trace.Trace) getIndexOptionsFunc {
+		startingFingerPrint := s.configFingerprint
+		tr.LazyPrintf("fingerprint: %s", startingFingerPrint)
+
+		first := true
+		return func(repos ...uint32) ([]indexOptionsItem, error) {
+			options, nextFingerPrint, err := s.getIndexOptionsREST(startingFingerPrint, repos...)
+			if err != nil {
+				first = false
+				s.configFingerprint = startingFingerPrint
+
+				return nil, err
+			}
+
+			if first {
+				first = false
+				s.configFingerprint = nextFingerPrint
+
+				tr.LazyPrintf("new fingerprint: %s", nextFingerPrint)
+			}
+
+			return options, nil
+		}
+	}
+
+	// If we enabled GRPC, use our gRPC client instead.
+	if s.shouldUseGRPCFunc() {
+		mkGetIndexOptionsFunc = func(tr trace.Trace) getIndexOptionsFunc {
+			startingFingerPrint := s.configFingerprintProto
+			tr.LazyPrintf("fingerprint: %s", startingFingerPrint.String())
+
+			first := true
+			return func(repos ...uint32) ([]indexOptionsItem, error) {
+				options, nextFingerPrint, err := s.getIndexOptionsGRPC(context.Background(), startingFingerPrint, repos)
+				if err != nil {
+					first = false
+					s.configFingerprintProto = startingFingerPrint
+
+					return nil, err
+				}
+
+				if first {
+					first = false
+					s.configFingerprintProto = nextFingerPrint
```

Comment:
It makes me a little bit uncomfortable that this closure hides mutation of `s`, especially since this can be seen as a form of cross-network control flow. Could we instead return the fingerprint from the closure and update `s` in the loop below?

Reply:
We talked about this a bit during our sync, and I was misunderstanding that the types of the two fingerprints are different. This makes it really difficult to abstract this, and it's probably not worth the effort.

Reply:
Discussed offline. The logic here is tricky since we need to properly maintain two different types of fingerprints (string or protobuf struct). 

Using a closure is the cleanest way we can think of to do this since introducing extra interfaces would needlessly complicate things .